packages/gatsby-theme-newrelic/src/components/Portal.js

```diff
@@ -0,0 +1,16 @@
+import { useState, useEffect } from 'react';
+import { createPortal } from 'react-dom';
+
+const Portal = ({ children }) => {
```

Comment:
Thinking out loud here, but perhaps we can turn this into a hook (something like `usePortal`) and expose the portal as well as some callbacks for when content gets loaded / removed from it? That way we don't need to do the `useEffect` business in the component. Not sure if that would even work.

Reply:
@zstix I can explain why the `useEffect` is important here.

Portals need an existing DOM element to attach itself to so that it knows where in the DOM to transport the UI. There are a couple ways of achieving this:

1) Modify the "template" HTML to add a DOM node with a specific ID that portals can attach themselves to. This requires an existing DOM node with that `id` attribute to act as a hook for the portal to attach itself to.

2) Create a DOM node on the fly when the component mounts and use that as the root element of the portal for the duration of the Portal. Once the portal is no longer needed, remove that root element from the DOM as part of the cleanup.

We've opted for 2) in this case to make it easy to wrap a bit of UI in a portal without needing the consumer of the component to manage that lifecycle. The component itself is a really nice shortcut for someone because they don't need to manage the whole `document.createElement`/`document.getElementById` lifecycle themselves. 

Knowing that we've opted for 2) this means we need some way of handling both the creation of the root DOM node for the portal, attaching it to the existing DOM, and the removal of that element from the DOM when the portal is no longer rendered. In this case `useEffect` handles this for us. We create a DOM node using the `useState` hook, then manage when its attached/removed from the DOM using the `useEffect` hook. Without the `useEffect` hook, there isn't a super great way to manage that lifecycle (unless there is something I'm unaware of, which is also very likely 😄 )

tl;dr; This component is a thin wrapper around the `createPortal` API so that we can wrap a bit of UI with this component instead of needing others to re-implement this logic over and over just to get something to render in the UI.

Does that make sense?

Reply:
can definitely pull it out into a custom hook
something like this?
```
const usePortal = (isOpen, onOpen, onReset) =>
  useEffect(() => {
    if (isOpen) onOpen();
    return onReset;
  }, [isOpen, onOpen, onReset]);

usePortal(
  isOpen,
  () => (document.body.style.overflow = 'hidden'),
  () => (document.body.style.overflow = null)
);

Reply:
I personally find that more complicated to use. The Portal is all about transporting some bit of the React tree to another part of the DOM. In this case, you have no way of specifying what elements are being transported and passed to `createPortal`

Current API:

```js
import { Portal } from '@newrelic/gatsby-theme-newrelic'

const MyComp = () => (
  <Portal>
    // this stuff will end up transported to another part of the DOM 
  </Portal>
)
```

Proposed:

```js
import { useState } from 'react'

const MyComp = () => {
  const [isOpen, setOpen] = useState(false)

  usePortal(isOpen, () => console.log('opened'), () => console.log('reset'))

  return createPortal(???, ???)
}
```